### PR TITLE
Move advisers admin screens

### PR DESCRIPTION
**Pull request by munckymagik**
Adds a button to the `/admin/firms/:id` page that accesses a wizard that guides you through moving advisers from one firm to another.

@neoeno @JiggyPete

### CODE REVIEW

File: app/views/admin/move_advisers/new.html.erb [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+    <ul class="t-errors bg-danger">
+      <% @form.errors.full_messages.each do |message| %>
+        <li><%= message %></li>
+      <% end %>
+    </ul>
+  <% end %>
+
+  <div class="form-group">
+    <%= f.collection_check_boxes(:adviser_ids, @form.from_firm.advisers.order(:reference_number),
+          :id, ->(adviser) { "(#{adviser.reference_number}) #{adviser.name}" }) do |b| %>
```
Comment (mconnell) [inline app/views/admin/move_advisers/new.html.erb]: The defining of a stabby proc here for the checkbox label is interesting. I've not seen that done before. You could add a method to adviser which does this `:ref_number_and_name` or something, but then I guess you're then polluting the model with decoration, so it's deciding what's the lesser of two evils. I can read and understand it as it stands so not suggesting you actually do that refactor.

**Comment by mconnell** [comment]
:+1:

File: app/controllers/admin/move_advisers_controller.rb [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+    before_action :assign_form
+
+    def new
+    end
+
+    def choose_to_firm
+      if @form.invalid?
+        render :new
+      end
+    end
```
Comment (neoeno) [inline app/controllers/admin/move_advisers_controller.rb]: `choose_destination_firm`, maybe?

File: app/controllers/admin/move_advisers_controller.rb [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+      @form = Admin::MoveAdvisersForm.new(safe_params)
+    end
+
+    def safe_params
+      params.permit(:id).merge(
+        params.fetch(:admin_move_advisers_form, {})
+          .permit(:to_firm_fca_number,
+                  :to_firm_id,
+                  adviser_ids: []))
+    end
```
Comment (neoeno) [inline app/controllers/admin/move_advisers_controller.rb]: Took me a while to understand the intention of this — might be worth a comment.

File: app/forms/admin/move_advisers_form.rb [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    def to_firm
+      Firm.find(to_firm_id)
+    end
+
+    def advisers_to_move
+      Adviser.where(id: adviser_ids).order(:reference_number)
+    end
+
+    def subsidiaries
+      Firm.registered.where(fca_number: to_firm_fca_number).sort_by { |f| f.registered_name.downcase }
```
Comment (neoeno) [inline app/forms/admin/move_advisers_form.rb]: Suggestion:

```
Firm.registered.where(fca_number: to_firm_fca_number).order('LOWER(registered_name)')
```

Could make it a scope of Firm if you prefer.

File: app/forms/admin/move_advisers_form.rb [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+    end
+
+    private
+
+    def to_firm_fca_number_exists
+      unless Firm.where(fca_number: to_firm_fca_number).count > 0
```
Comment (neoeno) [inline app/forms/admin/move_advisers_form.rb]: MAS styleguide prefers guard clauses I think

File: app/views/admin/firms/show.html.erb [hunk 6]
```diff
@@ -77,6 +77,11 @@
       <li><%= link_to adviser.name, admin_adviser_path(adviser) %>, added <%= adviser.created_at.to_s(:short) %></li>
     <% end %>
   </ul>
+
+  <div>
```
Comment (neoeno) [inline app/views/admin/firms/show.html.erb]: Some people recommend against using divs/spans with no attributes. Do we need this wrapper? & if so, is there a more meaningful element we could use here?

File: app/views/admin/move_advisers/choose_subsidiary.html.erb [hunk 7]
```diff
@@ -0,0 +1,27 @@
+<%= form_for(@form, url: confirm_admin_move_advisers_path, method: :get) do |f| %>
+  <h1>Choose subsidiary</h1>
+  <% if @form.errors.any? %>
+    <ul class="t-errors bg-danger">
+      <% @form.errors.full_messages.each do |message| %>
+        <li><%= message %></li>
+      <% end %>
+    </ul>
```
Comment (neoeno) [inline app/views/admin/move_advisers/choose_subsidiary.html.erb]: Do empty `ul` elements collapse? Might be able to avoid this conditional if so.

File: app/views/admin/move_advisers/choose_subsidiary.html.erb [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+  <% end %>
+
+  <div class="form-group">
+  <%= f.collection_radio_buttons(:to_firm_id, @form.subsidiaries, :id,
+      ->(firm) { "(#{firm.fca_number}) #{firm.registered_name}" }) do |b| %>
+    <div class="t-subsidiary radio">
+      <%= b.label(class: 't-subsidiary-label') { b.radio_button(class: 't-subsidiary-field') + b.text } %>
+    </div>
+  <% end %>
+  </div>
```
Comment (neoeno) [inline app/views/admin/move_advisers/choose_subsidiary.html.erb]: Indentation is off here.

File: app/views/admin/firms/show.html.erb (hunk 6 above)
Comment (benbarnett) [inline app/views/admin/firms/show.html.erb]: Agreed

File: app/views/admin/move_advisers/choose_subsidiary.html.erb [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+
+  <div class="t-hidden-fields form-group">
+    <% @form.adviser_ids.each do |adviser_id| %>
+      <%= f.hidden_field(:adviser_ids, multiple: true, value: adviser_id, class: 't-adviser') %>
+    <% end %>
+    <%= f.hidden_field(:to_firm_fca_number, value: @form.to_firm_fca_number, class: 't-to_firm_fca_number') %>
+    <%= f.submit('Next', class: 't-next') %>
+  </div>
```
Comment (neoeno) [inline app/views/admin/move_advisers/choose_subsidiary.html.erb]: Does this submit button belong in this element? `t-hidden-fields` seems incorrect if so.

File: app/views/admin/move_advisers/choose_to_firm.html.erb [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+    <%= f.text_field :to_firm_fca_number, class: 't-to_firm_fca_number' %>
+  </div>
+
+  <div class="t-hidden-fields form-group">
+    <% @form.adviser_ids.each do |adviser_id| %>
+      <%= f.hidden_field(:adviser_ids, multiple: true, value: adviser_id, class: 't-adviser') %>
+    <% end %>
+    <%= f.submit('Next', class: 't-next') %>
+  </div>
+<% end %>
```
Comment (neoeno) [inline app/views/admin/move_advisers/choose_to_firm.html.erb]: As in `choose_subsidiary.html.erb`

File: app/views/admin/move_advisers/confirm.html.erb [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+
+<p>Please confirm you want to:</p>
+<dl>
+  <dt>Move:</dt>
+  <% @form.advisers_to_move.each do |adviser| %>
+    <dd class="t-adviser">(<%= adviser.reference_number %>) <%= adviser.name %></dd>
+  <% end %>
+  <dt>From:</dt><dd><span class="t-from_firm">(<%= @form.from_firm.fca_number %>)
+    <%= @form.from_firm.registered_name %></span></dd>
+  <dt>To:</dt><dd><span class="t-to_firm">(<%= @form.to_firm.fca_number %>)
```
Comment (neoeno) [inline app/views/admin/move_advisers/confirm.html.erb]: There's no hard and fast rule on formatting test classes I think, but generally I'd expect to see hyphens than underscores. Nitpicky I know!

File: app/views/admin/move_advisers/confirm.html.erb [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+<%= form_for(@form, url: move_admin_move_advisers_path) do |f| %>
+  <div class="t-hidden-fields">
+    <% @form.adviser_ids.each do |adviser_id| %>
+      <%= f.hidden_field(:adviser_ids, multiple: true, value: adviser_id, class: 't-adviser') %>
+    <% end %>
+    <%= f.hidden_field(:to_firm_id, class: 't-to_firm_id') %>
+    <%= f.submit('Move', class: 't-move') %>
+    <%= link_to 'Cancel', admin_firm_path(@form.from_firm), :method => :get %>
+  </div>
+<% end %>
```
Comment (neoeno) [inline app/views/admin/move_advisers/confirm.html.erb]: Hidden fields stuff again. Or am I missing something?

File: spec/features/admin/move_advisers_spec.rb [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+    then_the_adviser_is_moved
+  end
+
+  scenario 'When no advisers are selected to move' do
+    given_i_do_not_specify_any_advisers_to_move
+    then_i_am_told_about_the_problem_on(choose_firm_page)
+  end
+
+  scenario 'When the to-firm does not exist' do
+    given_i_supply_a_non_existent_fca_number('DOES_NOT_EXIST')
```
Comment (neoeno) [inline spec/features/admin/move_advisers_spec.rb]: Can we drop this argument? It seems implicit in your name here.

File: spec/features/admin/move_advisers_spec.rb [hunk 14]
```diff
@@ -0,0 +1,113 @@
...
+    then_i_am_told_about_the_problem_on(choose_subsidiary_page)
```
Comment (neoeno) [inline spec/features/admin/move_advisers_spec.rb]: Liking the pattern here, though.

File: app/views/admin/move_advisers/choose_subsidiary.html.erb (hunk 7 above)
Comment (benbarnett) [inline app/views/admin/move_advisers/choose_subsidiary.html.erb]: It depends on their styles, probably best to leave the conditional to avoid stray elements

**Comment by neoeno** [comment]
Nothing major, just a lot of code to find minor things in.

Nice work, @munckymagik @JiggyPete :+1:

File: app/views/admin/move_advisers/confirm.html.erb (hunk 11 above)
Comment (benbarnett) [inline app/views/admin/move_advisers/confirm.html.erb]: Agreed also :) We do use `t-` to prefix test classes though, so that's cool.

Comment (munckymagik) [inline app/views/admin/move_advisers/confirm.html.erb]: If I do it, I'd like to do it in a one-er. But, I'm tempted to do this in a separate PR because there are other instances of the underscore in the existing code. wdyt?
